Sources/NIOCore/AsyncSequences/NIOAsyncSequenceProducer.swift

```diff
@@ -0,0 +1,1064 @@
+                }
+
+                // We have an element and can resume the continuation
+
+                let shouldProduceMore = backPressureStrategy.didYield(bufferDepth: buffer.count)
+                self.state = .streaming(
+                    backPressureStrategy: backPressureStrategy,
+                    buffer: buffer,
+                    continuation: nil, // Setting this to nil since we are resuming the continuation
+                    hasOutstandingDemand: shouldProduceMore,
+                    iteratorInitialized: iteratorInitialized
+                )
+
+                return .init(shouldProduceMore: shouldProduceMore, continuationAndElement: (continuation, element))
+
+            case .streaming(var backPressureStrategy, var buffer, continuation: .none, _, let iteratorInitialized):
+                self.state = .modifying
+
+                buffer.append(contentsOf: sequence)
+                let shouldProduceMore = backPressureStrategy.didYield(bufferDepth: buffer.count)
+
+                self.state = .streaming(
+                    backPressureStrategy: backPressureStrategy,
+                    buffer: buffer,
+                    continuation: nil,
+                    hasOutstandingDemand: shouldProduceMore,
+                    iteratorInitialized: iteratorInitialized
+                )
+
+                return .init(shouldProduceMore: shouldProduceMore)
+
+            case .sourceFinished, .finished:
+                // If the source has finished we are dropping the elements.
+                return .returnDropped
+
+            case .modifying:
+                preconditionFailure("Invalid state")
+            }
+        }
+
+        /// Actions returned by `finish()`.
+        @usableFromInline
+        enum FinishAction {
+            /// Indicates that ``NIOAsyncSequenceProducerDelegate/didTerminate()`` should be called.
+            case callDidTerminate
+            /// Indicates that the continuation should be resumed with `nil` and
+            /// that ``NIOAsyncSequenceProducerDelegate/didTerminate()`` should be called.
+            case resumeContinuationWithNilAndCallDidTerminate(CheckedContinuation<Element?, Never>)
+            /// Indicates that nothing should be done.
+            case none
+        }
+
+        @inlinable
+        mutating func finish() -> FinishAction {
+            switch self.state {
+            case .initial:
+                // Nothing was yielded nor did anybody call next
+                // This means we can just transition to finished
+                self.state = .finished
+
+                return .callDidTerminate
+
+            case .streaming(_, let buffer, .some(let continuation), _, _):
+                // We have a continuation, this means our buffer must be empty
+                // Furthermore, we can now transition to finished
+                // and resume the continuation with `nil`
+                precondition(buffer.isEmpty)
```

Comment:
Please add a short message; those are so annoying to debug when they'd happen and you don't get a backtrace or even a message; always nicer to at least have a short string to grep for